eZ/Publish/SPI/Search/Content/IndexerGateway.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\SPI\Search\Content;
+
+use DateTime;
+use Doctrine\DBAL\Driver\ResultStatement;
+use Generator;
+
+/**
+ * @internal
+ */
+interface IndexerGateway
+{
+    /**
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function getStatementContentSince(DateTime $since, bool $count = false): ResultStatement;
+
+    /**
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function getStatementSubtree(string $locationPath, bool $count = false): ResultStatement;
+
+    /**
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function getStatementContentAll(bool $count = false): ResultStatement;
+
+    /**
+     * @return \Generator a list of Content IDs for each iteration
+     */
+    public function fetchIteration(ResultStatement $stmt, int $iterationCount): Generator;
```

Comment:
We could type-hint against an `Iterator` or a `Traversable` instead of a `Generator`. 

Our implementation returns a `Generator`, but we should be fine even with `iterable`?

Also, we should also document that we expect that it returns `Generator<int[]>`.

Reply:
> We could type-hint against an `Iterator` or a `Traversable` instead of a `Generator`.
> 
> Our implementation returns a `Generator`, but we should be fine even with `iterable`?
> 
> Also, we should also document that we expect that it returns `Generator<int[]>`.

Implementation uses
```php
 $generator->next();
 $generator->valid();
```
so we can use `\Iterator` if you insist.

Reply:
I definitely do not insist on this one. `Generator` comes with a few quirks of it's own, I'm more than fine with leaving it in this particular case. Especially since, as you noted, we use specific `Iterator` methods.

So ideally just add the phpdoc annotation about what specific data this `Generator` returns (`int[]`) and I'm more than happy :smile:

Reply:
@Steveb-p unfortunately code analysis does not understand that:
![image](https://user-images.githubusercontent.com/7099219/113857240-91b53480-97a2-11eb-9dc1-ab4d393c4817.png)

Reply:
One day, PHPStorm will respect those...

Surely...

You can add `@phpstan-return` that would be recognized by this particular tool, but... at this point I'd say it's irrelevant, thanks for bearing with me :smile: